sdk/nodejs/runtime/invoke.ts

```diff
@@ -100,3 +127,135 @@ export async function invoke(tok: string, props: Inputs, opts?: InvokeOptions):
         done();
     }
 }
+
+function invokeSync(tok: string, props: any, opts: InvokeOptions = {}): Promise<any> {
+    const label = `Invoking function: tok=${tok} synchronously`;
+    log.debug(label +
+        excessiveDebugOutput ? `, props=${JSON.stringify(props)}` : ``);
+
+    const obj = gstruct.Struct.fromJavaScript(serializePropertiesSync(props));
+    log.debug(`Invoke RPC prepared: tok=${tok}` + excessiveDebugOutput ? `, obj=${JSON.stringify(obj)}` : ``);
+
+    // Fetch the sync monitor and make an RPC request.
+    const syncInvokes = getSyncInvokes();
+
+    const provider = getProvider(tok, opts);
+
+    let providerRef: string | undefined;
+    if (ProviderRef.isInstance(provider)) {
+        providerRef = provider.getValue();
+    }
+    else if (Resource.isInstance(provider)) {
+        // TODO(cyrusn): issue warning here that we are synchronously blocking an rpc call.
+        providerRef = utils.promiseResult(ProviderRef.get(provider)).getValue();
+    }
+
+    const req = new providerproto.InvokeRequest();
+    req.setTok(tok);
+    req.setArgs(obj);
+    req.setProvider(providerRef);
+    req.setVersion(opts.version || "");
+
+    // Encode the request.
+    const reqBytes = Buffer.from(req.serializeBinary());
+
+    // Write the request length.
+    const reqLen = Buffer.alloc(4);
+    reqLen.writeUInt32BE(reqBytes.length, 0);
+    fs.writeSync(syncInvokes.requests, reqLen);
+    fs.writeSync(syncInvokes.requests, reqBytes);
+
+    // Read the response.
+    const respLenBytes = Buffer.alloc(4);
+    fs.readSync(syncInvokes.responses, respLenBytes, 0, 4, null);
+    const respLen = respLenBytes.readUInt32BE(0);
+    const respBytes = Buffer.alloc(respLen);
+    fs.readSync(syncInvokes.responses, respBytes, 0, respLen, null);
+
+    // Decode the response.
+    const resp = providerproto.InvokeResponse.deserializeBinary(new Uint8Array(respBytes));
+
+    // If there were failures, propagate them.
+    processPotentialFailures(tok, resp);
+
+    // Finally propagate any other properties that were given to us as outputs.
+    const resultValue = deserializeProperties(resp.getReturn());
+
+    // Expose the properties of the actual result of invoke directly on the promise itself.
+    const promise = Promise.resolve(resultValue);
+    Object.assign(promise, resultValue);
+
+    return promise;
```

Comment:
this is to preserve the shape of `invoke` entrypoint we've already shipped.  We may have existing code that calls into a dataSource like so: `aws.x.getY().then(r => r.z)`.  We need to make sure what we return is still a Promise so that htat will work.  However, we lift all the properties onto the promise as well (just like we used to) so that `aws.x.getY().z` still works as well.

Reply:
even though there is a promise here, this function is still entire synchronous.